In [1]:
# Importing the required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
pd.set_option('display.max_columns', None)

In [3]:
# Loading the customer golden repository
gr = pd.read_csv('Customer_Golden_Repository.csv')
print(gr.shape)
gr.head(3)

/usr/local/src/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(1897608, 7)


Unnamed: 0                VIN           NAME  MOBILE_NUM  \
0           0  MALBB51RLDM610345    Abhay Rasal  7020192479   
1           1  MALFC81BLLM153785    abhay singh  8090849984   
2           2  MALAM51BLFM622594  pragya pandey  8251829946   

                EMAIL_ADDR SOURCE     DELIVERY_STATUS  
0    abhayrasal3@gmail.com  Hcare  REJECTED-MULTIPART  
1  abhayrihu2012@gmail.com  Hcare             DELIVRD  
2      abhayrits@gmail.com  Hcare             DELIVRD

In [4]:
# Loading the RO Billing Data without taxi VINs and Refurbished RO Orders
ro_data = pd.read_csv('RO_Data_without_taxi_and_rf_orders.csv')
print(ro_data.shape)
ro_data.head(3)

(56069, 44)


/usr/local/src/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (23,41) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unnamed: 0                VIN                 NAME           MODEL  \
0           0  KMHSH81XLAU665847       D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516  PREM DEVA PRASATH.D  SANTA FE (CMi)   
2           2  MALA151ALCM047539           KUMARA N K             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U           kumaran@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2       3rd avenue 5th main teachers colony      KA01MH6810     20120224.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            S8413  C2017010036                   NaN              NaN   

  MODEL_CODE  CITY_NAME   STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM     MUMBAI  Maharashtra        400081          W1202   
1         BM   AVANASHI    Tamilnadu        641653          S4407   
2         IA  BANGALORE    Karnataka        560034          S8413   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        S8214        RR       1202  2021-09-04 14:28:45  R202103063   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8867907973        NaN           NaN    36210                 NaN   

  CUST_GENDER  CUST_AGE PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN        NaN                         NaN   
1           M       NaN        NaN                         NaN   
2           M       NaN        NaN                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0                  7141   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                    243.40                     165.2   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-04 16:31:35    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                 P1   

                                        Deffered Job         SA_NAME Taxi_flag  
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...    KISHOR KADAM        No  
1                                    Wheel Alignment      Karthik K.        No  
2                                      PAID  SERVICE  GURU PRASAD .G        No

In [5]:
# Checking the data types of variables in RO Table
ro_data.dtypes

Unnamed: 0                      int64
VIN                            object
NAME                           object
MODEL                          object
FUEL_TYPE                      object
EMAIL_ID                       object
ADDRESS                        object
REGISTRATION_NO                object
PURCHASE_DATE                 float64
PREFERRED_DEALER               object
REPM_CUST_NO                   object
SPEC_AUTO_TRANS_FALG          float64
SPEC_ENGN_CPCTY               float64
MODEL_CODE                     object
CITY_NAME                      object
STATE_NAME                     object
CUST_PIN_CODE                  object
SERVICE_DLR_NO                 object
SALES_DLR_NO                   object
WORK_TYPE                      object
VRNT_CODE                       int64
RO_DTIME                       object
RO_NO                          object
MOBILE_NO                      object
OFFICE_NO                     float64
RESIDENCE_NO                  float64
MILEAGE     

In [6]:
# Checking if there are any taxi VINs in the ro data
ro_data['Taxi_flag'].value_counts()

No    56069
Name: Taxi_flag, dtype: int64

In [7]:
# Getting unique phone no values for each vin
ro_data.sort_values(['VIN', 'RO_DTIME'], inplace=True)
ro_data.drop_duplicates(['VIN', 'RO_DTIME'], inplace=True)

In [8]:
ro_data.head(2)

Unnamed: 0                VIN                 NAME           MODEL  \
0           0  KMHSH81XLAU665847       D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516  PREM DEVA PRASATH.D  SANTA FE (CMi)   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   

  MODEL_CODE CITY_NAME   STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO SALES_DLR_NO  \
0         BM    MUMBAI  Maharashtra        400081          W1202        W1201   
1         BM  AVANASHI    Tamilnadu        641653          S4407        S1100   

  WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO   MOBILE_NO  \
0        RR        963  2021-08-25 13:52:04  R202116390  9821321604   
1        RR        963  2021-09-29 16:16:28  R202104381  9003429027   

   OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH CUST_GENDER  CUST_AGE  \
0        NaN           NaN    89063                 NaN           M       NaN   
1        NaN           NaN   149036                 NaN           M       NaN   

  PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT  \
0        NaN                         NaN                  NaN             0.0   
1        NaN                         NaN                  NaN          2000.0   

  REPM_SVC_ADVSR_EMP_NO  REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  TEC5                  50214.94                    6195.0   
1                 KM015                      0.00                    1888.0   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   

                                        Deffered Job       SA_NAME Taxi_flag  
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No  
1                                    Wheel Alignment    Karthik K.        No

In [9]:
# Checking the shape of the RO data
ro_data.shape

(55958, 44)

In [10]:
56069-55958

111

There were 111 Repeated VINs which have been removed.

In [11]:
# Checking the count of unique vins and phone no
print('Unique VINs:', ro_data['VIN'].nunique())
print('Unique Mobile no:', ro_data['MOBILE_NO'].nunique())

Unique VINs: 55958
Unique Mobile no: 55645


In [12]:
55958-55645

313

Around 313 are repeated phone numbers in Unique VINs.

In [13]:
# Getting a list of VIN and mobile no to merge the dataset
latest_ro_mobile_no = ro_data[['VIN', 'MOBILE_NO']]
print(latest_ro_mobile_no.shape)
latest_ro_mobile_no.head(3)

(55958, 2)


VIN   MOBILE_NO
0      KMHSH81XLAU665847  9821321604
1      KMHSH81XLBU741516  9003429027
46028  MALA151ALCM039126  8600031276

In [14]:
# Taking only VIN and Mobile no from Golden Repository
gr_vin_mobile = gr[['VIN', 'MOBILE_NUM']]
print(gr_vin_mobile.shape)
gr_vin_mobile.head(3)

(1897608, 2)


VIN  MOBILE_NUM
0  MALBB51RLDM610345  7020192479
1  MALFC81BLLM153785  8090849984
2  MALAM51BLFM622594  8251829946

In [15]:
# Merging RO mobile and Golden Repository Mobile no
first_merged = pd.merge(latest_ro_mobile_no, gr_vin_mobile,how="left", on=["VIN"])
print(first_merged.shape)
first_merged.head()

(55958, 3)


VIN    MOBILE_NO    MOBILE_NUM
0  KMHSH81XLAU665847   9821321604  9.821322e+09
1  KMHSH81XLBU741516   9003429027  9.003429e+09
2  MALA151ALCM039126   8600031276           NaN
3  MALA151ALCM042203  9.83622e+09  8.335822e+09
4  MALA151ALCM044051   6200979987  6.200980e+09

In [16]:
# Checking how many unique VINs are there in the data
first_merged['VIN'].nunique()

55958

In [17]:
# Checking how many unique Mobile no are there in the RO data
first_merged['MOBILE_NO'].nunique()

55645

In [18]:
# Renaming the columns in the data set
first_merged.rename({'MOBILE_NO': 'RO_Mobile_No', 'MOBILE_NUM': 'GR_Mobile_no'}, axis=1, inplace=True)

In [19]:
# Checking whether the columns have been renamed or not
first_merged.head(3)

VIN RO_Mobile_No  GR_Mobile_no
0  KMHSH81XLAU665847   9821321604  9.821322e+09
1  KMHSH81XLBU741516   9003429027  9.003429e+09
2  MALA151ALCM039126   8600031276           NaN

In [20]:
first_merged.dtypes

VIN              object
RO_Mobile_No     object
GR_Mobile_no    float64
dtype: object

In [21]:
# Cleaning phone numbers in the data frame = df[df['reason'].str.len() <= 10]
clean_phone_no = first_merged[first_merged['RO_Mobile_No'].str.len() == 10]
print(clean_phone_no.shape)
clean_phone_no.head(3)

(32672, 3)


VIN RO_Mobile_No  GR_Mobile_no
0  KMHSH81XLAU665847   9821321604  9.821322e+09
1  KMHSH81XLBU741516   9003429027  9.003429e+09
2  MALA151ALCM039126   8600031276           NaN

In [22]:
clean_phone_no1 = clean_phone_no[clean_phone_no['RO_Mobile_No'] != 'SATISH KUM']

In [23]:
# Changing the dtype of RO mobile no
clean_phone_no1['RO_Mobile_No'] = clean_phone_no1.RO_Mobile_No.astype(float)

/usr/local/src/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [24]:
# Adding a column as final Mobile no
clean_phone_no1['Verified_mob_no'] = np.where(clean_phone_no1['RO_Mobile_No'] == clean_phone_no1['GR_Mobile_no'], clean_phone_no1['GR_Mobile_no'],  clean_phone_no1['RO_Mobile_No'])
print(clean_phone_no1.shape)
clean_phone_no1.head(3)

(32671, 4)


/usr/local/src/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


VIN  RO_Mobile_No  GR_Mobile_no  Verified_mob_no
0  KMHSH81XLAU665847  9.821322e+09  9.821322e+09     9.821322e+09
1  KMHSH81XLBU741516  9.003429e+09  9.003429e+09     9.003429e+09
2  MALA151ALCM039126  8.600031e+09           NaN     8.600031e+09

In [25]:
clean_phone_no1.isnull().sum()

VIN                    0
RO_Mobile_No           0
GR_Mobile_no       11022
Verified_mob_no        0
dtype: int64

In [26]:
# Now loading HMM mobile no data
# Then replacing the verified mobile no with Connectivity mobile no

# Loading the HMM Data
hmm = pd.read_excel('HMM_Members_Phone_no.xltm')
print(hmm.shape)
hmm.head(3)

(553539, 2)


VIN       Phone
0  MALB341CYMM071370  8187093036
1  MALBH512TMM062250  9948234482
2  MALCU41ULFM200758  9553620354

In [27]:
# Renaming the columns
hmm.rename({'Phone': 'HMM_Mobile_no'}, axis = 1, inplace = True)
hmm.head(2)

VIN  HMM_Mobile_no
0  MALB341CYMM071370     8187093036
1  MALBH512TMM062250     9948234482

In [28]:
# Adding HMM contacts to verified mobile no
hmm_merged = pd.merge(clean_phone_no1, hmm,how = "left", on = ["VIN"])
print(hmm_merged.shape)
hmm_merged.head()

(32671, 5)


VIN  RO_Mobile_No  GR_Mobile_no  Verified_mob_no  \
0  KMHSH81XLAU665847  9.821322e+09  9.821322e+09     9.821322e+09   
1  KMHSH81XLBU741516  9.003429e+09  9.003429e+09     9.003429e+09   
2  MALA151ALCM039126  8.600031e+09           NaN     8.600031e+09   
3  MALA151ALCM044051  6.200980e+09  6.200980e+09     6.200980e+09   
4  MALA151ALCM047539  8.867908e+09  8.310257e+09     8.867908e+09   

   HMM_Mobile_no  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

In [29]:
hmm_merged.isnull().sum()

VIN                    0
RO_Mobile_No           0
GR_Mobile_no       11022
Verified_mob_no        0
HMM_Mobile_no      26415
dtype: int64

In [31]:
32671-26415

6256

Out of 32671 records in RO Data only 6256 have been regitered on HMM. So only for these 6256 records HMM phone no will be taken as Verified mobile no.

In [32]:
# Updating the mobile no for these 12026 records
hmm_merged['Verified_mob_no'] = np.where(hmm_merged['HMM_Mobile_no'].notnull(), hmm_merged['HMM_Mobile_no'], hmm_merged['Verified_mob_no'] )
print(hmm_merged.shape)
hmm_merged.head(3)

(32671, 5)


VIN  RO_Mobile_No  GR_Mobile_no  Verified_mob_no  \
0  KMHSH81XLAU665847  9.821322e+09  9.821322e+09     9.821322e+09   
1  KMHSH81XLBU741516  9.003429e+09  9.003429e+09     9.003429e+09   
2  MALA151ALCM039126  8.600031e+09           NaN     8.600031e+09   

   HMM_Mobile_no  
0            NaN  
1            NaN  
2            NaN

In [33]:
# Loading the connectivity data
connectivity = pd.read_csv('Connectivity_Registered_Mobile_User_data.csv')
print(connectivity.shape)
connectivity.head(3)

(173870, 5)


Unnamed: 0             vin_id    enroll_status                prov_last_dt  \
0           0  MALFE81CLKM000041  FULL ENROLLMENT  2019-08-29 10:49:22.780788   
1           1  KMHFC81AVK1000115  FULL ENROLLMENT  2021-03-12 11:15:59.657030   
2           2  MALBK512LMM035269  FULL ENROLLMENT  2021-03-28 08:26:37.860336   

       phone_no  
0  1.112346e+07  
1  1.112346e+07  
2  9.090415e+09

In [34]:
# Considering both types of enrollment status for verified phone no
connectivity['enroll_status'].value_counts()

FULL ENROLLMENT          155755
ÍN VEHICLE ENROLLMENT     18115
Name: enroll_status, dtype: int64

In [35]:
# Renaming the phone no column
connectivity.rename({'phone_no': 'Bluelink_Mobile_no'}, axis=1, inplace=True)
connectivity.head(3)

Unnamed: 0             vin_id    enroll_status                prov_last_dt  \
0           0  MALFE81CLKM000041  FULL ENROLLMENT  2019-08-29 10:49:22.780788   
1           1  KMHFC81AVK1000115  FULL ENROLLMENT  2021-03-12 11:15:59.657030   
2           2  MALBK512LMM035269  FULL ENROLLMENT  2021-03-28 08:26:37.860336   

   Bluelink_Mobile_no  
0        1.112346e+07  
1        1.112346e+07  
2        9.090415e+09

In [36]:
# Taking only vin and mobile no from connectivity data frame
conn = connectivity[['vin_id', 'Bluelink_Mobile_no']]
print(conn.shape)
conn.head(3)

(173870, 2)


vin_id  Bluelink_Mobile_no
0  MALFE81CLKM000041        1.112346e+07
1  KMHFC81AVK1000115        1.112346e+07
2  MALBK512LMM035269        9.090415e+09

In [37]:
# Merging the mobile no with verified mobile no list for connectivity numbers
conn_merged = pd.merge(hmm_merged, conn,how = "left", left_on = ["VIN"], right_on = ['vin_id'])
print(conn_merged.shape)
conn_merged.head()

(32671, 7)


VIN  RO_Mobile_No  GR_Mobile_no  Verified_mob_no  \
0  KMHSH81XLAU665847  9.821322e+09  9.821322e+09     9.821322e+09   
1  KMHSH81XLBU741516  9.003429e+09  9.003429e+09     9.003429e+09   
2  MALA151ALCM039126  8.600031e+09           NaN     8.600031e+09   
3  MALA151ALCM044051  6.200980e+09  6.200980e+09     6.200980e+09   
4  MALA151ALCM047539  8.867908e+09  8.310257e+09     8.867908e+09   

   HMM_Mobile_no vin_id  Bluelink_Mobile_no  
0            NaN    NaN                 NaN  
1            NaN    NaN                 NaN  
2            NaN    NaN                 NaN  
3            NaN    NaN                 NaN  
4            NaN    NaN                 NaN

In [38]:
# Checking for null values
conn_merged.isnull().sum()

VIN                       0
RO_Mobile_No              0
GR_Mobile_no          11022
Verified_mob_no           0
HMM_Mobile_no         26415
vin_id                30783
Bluelink_Mobile_no    30783
dtype: int64

In [ ]:
57112 - 52026

In [39]:
# Checking whether the values have been updated correctly or not
conn_merged['Verified_mob_no'] = np.where(conn_merged['Bluelink_Mobile_no'].notnull(), conn_merged['Bluelink_Mobile_no'], conn_merged['Verified_mob_no'] )
print(conn_merged.shape)
conn_merged.tail(10)

(32671, 7)


VIN  RO_Mobile_No  GR_Mobile_no  Verified_mob_no  \
32661  MALSU81XMEM010586  9.921554e+09           NaN     9.921554e+09   
32662  MALSU81XMFM012124  9.904949e+09           NaN     9.904949e+09   
32663  MALSU81XMFM012626  9.825700e+09  9.904729e+09     9.904729e+09   
32664  MALSU81XMFM012821  9.610524e+09  9.610524e+09     9.610524e+09   
32665  MALSW81XSEM010630  9.490973e+09           NaN     9.490973e+09   
32666  MALSW81XSEM010668  9.444073e+09  9.444073e+09     9.444073e+09   
32667  MALSW81XSEM010679  9.019889e+09  9.845367e+09     9.019889e+09   
32668  MALSW81XSEM010744  9.973150e+09           NaN     9.973150e+09   
32669  MALSW81XSFM011900  9.487998e+09  9.487998e+09     9.487998e+09   
32670  MALSW81XSFM012509  9.951046e+09  9.951046e+09     9.951046e+09   

       HMM_Mobile_no vin_id  Bluelink_Mobile_no  
32661            NaN    NaN                 NaN  
32662            NaN    NaN                 NaN  
32663   9.904729e+09    NaN                 NaN  
32664            NaN    NaN                 NaN  
32665            NaN    NaN                 NaN  
32666            NaN    NaN                 NaN  
32667            NaN    NaN                 NaN  
32668            NaN    NaN                 NaN  
32669            NaN    NaN                 NaN  
32670   9.951046e+09    NaN                 NaN

#### Creating unique customer ID on RO Billing data by combining phone + VIN

In [40]:
# Checking the ro data
print(ro_data.shape)
ro_data.head(3)

(55958, 44)


Unnamed: 0                VIN                  NAME           MODEL  \
0               0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1               1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
46028       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

      FUEL_TYPE                    EMAIL_ID  \
0             D     dkumars9@rediffmail.com   
1             D  dpremdevaprasath@gmail.com   
46028         U      lalipcodiiya@gmail.com   

                                        ADDRESS REGISTRATION_NO  \
0                                        MUMBAI      MH04EQ5769   
1      4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700   
46028                           KASTURBA NAGAR        MP43C5146   

       PURCHASE_DATE PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  \
0         20101203.0            W1202  C2017090133                   NaN   
1         20110608.0            S4407  C2011060072                   NaN   
46028     20120331.0            W5237  C2019052081                   NaN   

       SPEC_ENGN_CPCTY MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE  \
0                  NaN         BM    MUMBAI     Maharashtra        400081   
1                  NaN         BM  AVANASHI       Tamilnadu        641653   
46028              NaN         IA    Ratlam  MADHYA PRADESH        457001   

      SERVICE_DLR_NO SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME  \
0              W1202        W1201        RR        963  2021-08-25 13:52:04   
1              S4407        S1100        RR        963  2021-09-29 16:16:28   
46028          W5237        W5208        PS       1178  2021-09-13 11:23:59   

            RO_NO   MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  \
0      R202116390  9821321604        NaN           NaN    89063   
1      R202104381  9003429027        NaN           NaN   149036   
46028  R202103318  8600031276        NaN           NaN    18177   

       CUST_DATE_OF_BIRTH CUST_GENDER  CUST_AGE   PROFESSION  \
0                     NaN           M       NaN          NaN   
1                     NaN           M       NaN          NaN   
46028                 NaN           M       NaN  Businessmen   

       SERVICE_ADVISOR_DATE_BIRTH  SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT  \
0                             NaN                  NaN             0.0   
1                             NaN                  NaN          2000.0   
46028                         NaN                  NaN             0.0   

      REPM_SVC_ADVSR_EMP_NO  REPM_BILL_TOTAL_PART_AMT  \
0                      TEC5                  50214.94   
1                     KM015                      0.00   
46028            EW52370040                   2376.98   

       REPM_BILL_TOTAL_LABR_AMT  REPM_BILL_TOTAL_OTHR_AMT  \
0                        6195.0                         0   
1                        1888.0                         0   
46028                    3380.7                         0   

           REPM_CLSE_DTIME INVC_TYPE                           DEF_CODE  \
0      2021-08-25 16:23:04    Normal  121,A05,B07,BD154,BR001,CL99,S002   
1      2021-09-29 17:15:47    Normal                           AW25,Z01   
46028  2021-09-13 17:35:48    Normal                                Z01   

                                            Deffered Job       SA_NAME  \
0      CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM   
1                                        Wheel Alignment    Karthik K.   
46028                                                NaN  IBRAHIM KHAN   

      Taxi_flag  
0            No  
1            No  
46028        No

In [41]:
# Adding the phone number column in RO Data
updated_ro_data = pd.merge(ro_data, conn_merged[['VIN','Verified_mob_no']],how = "left", left_on = ["VIN"], right_on = ['VIN'])
print(updated_ro_data.shape)
updated_ro_data.head()

(55958, 45)


Unnamed: 0                VIN                    NAME           MODEL  \
0           0  KMHSH81XLAU665847          D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516     PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126    JASWANT SINGH DODIYA             EON   
3       53917  MALA151ALCM042203  RAKESH KUMAR SATYARTHI             EON   
4       41178  MALA151ALCM044051     CHANDAN KUMAR SINGH             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   
3         U            RAKESH@GMAIL.COM   
4         U            sadush@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   
3                                  SILIGURI      JH05AN8594     20120503.0   
4                        CHANDWA MORE ARRAH      BR01BJ9344     20120314.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   
3            E1225  C2019100075                   NaN              NaN   
4            E6817  C2019113085                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   
3         IA  Siliguri     WEST BENGAL        734001          E1225   
4         IA     Arrah           BIHAR        802312          E6817   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   
3        E6201        PS       1178  2021-02-06 11:40:59  R202100793   
4        E6200        RR       1178  2021-05-01 13:57:41  R202100408   

     MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0   9821321604        NaN           NaN    89063                 NaN   
1   9003429027        NaN           NaN   149036                 NaN   
2   8600031276        NaN           NaN    18177                 NaN   
3  9.83622e+09        NaN           NaN    50405                 NaN   
4   6200979987        NaN           NaN    55100                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   
3           M       NaN          NaN                         NaN   
4         NaN       NaN          NaN                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   
3                  NaN             0.0            EE12250045   
4                  NaN             0.0            EE68170019   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                   6195.00   
1                      0.00                   1888.00   
2                   2376.98                   3380.70   
3                   5001.44                   3430.85   
4

In [42]:
# Creating a unique customer ID using phone + VIN
updated_ro_data.dtypes

Unnamed: 0                      int64
VIN                            object
NAME                           object
MODEL                          object
FUEL_TYPE                      object
EMAIL_ID                       object
ADDRESS                        object
REGISTRATION_NO                object
PURCHASE_DATE                 float64
PREFERRED_DEALER               object
REPM_CUST_NO                   object
SPEC_AUTO_TRANS_FALG          float64
SPEC_ENGN_CPCTY               float64
MODEL_CODE                     object
CITY_NAME                      object
STATE_NAME                     object
CUST_PIN_CODE                  object
SERVICE_DLR_NO                 object
SALES_DLR_NO                   object
WORK_TYPE                      object
VRNT_CODE                       int64
RO_DTIME                       object
RO_NO                          object
MOBILE_NO                      object
OFFICE_NO                     float64
RESIDENCE_NO                  float64
MILEAGE     

In [43]:
# Changing the type of Verified mobile no from Float to string
updated_ro_data['Verified_mob_no'] = updated_ro_data['Verified_mob_no'].astype(str)

In [44]:
# Checking the data type again
updated_ro_data.dtypes

Unnamed: 0                      int64
VIN                            object
NAME                           object
MODEL                          object
FUEL_TYPE                      object
EMAIL_ID                       object
ADDRESS                        object
REGISTRATION_NO                object
PURCHASE_DATE                 float64
PREFERRED_DEALER               object
REPM_CUST_NO                   object
SPEC_AUTO_TRANS_FALG          float64
SPEC_ENGN_CPCTY               float64
MODEL_CODE                     object
CITY_NAME                      object
STATE_NAME                     object
CUST_PIN_CODE                  object
SERVICE_DLR_NO                 object
SALES_DLR_NO                   object
WORK_TYPE                      object
VRNT_CODE                       int64
RO_DTIME                       object
RO_NO                          object
MOBILE_NO                      object
OFFICE_NO                     float64
RESIDENCE_NO                  float64
MILEAGE     

In [45]:
updated_ro_data.head(3)

Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no  
0    9821321604.0  
1    9003429027.0  
2    8600031276.0

In [46]:
# Extracting last 4 digits of VIN and storing in another column
updated_ro_data['VIN_end_digits'] = updated_ro_data['VIN'].str[-4:]
updated_ro_data['Mob_end_digits'] = updated_ro_data['Verified_mob_no'].str[:4]
print(updated_ro_data.shape)
updated_ro_data.head(3)

(55958, 47)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits  
0    9821321604.0           5847           9821  
1    9003429027.0           1516           9003  
2    8600031276.0           9126           8600

In [47]:
# Creating unique customer ID by merging the last VIN digits and first 4 verified mobile no digits
updated_ro_data['unique_customer_id'] = updated_ro_data['VIN_end_digits'] + updated_ro_data['Mob_end_digits']
print(updated_ro_data.shape)
updated_ro_data.head(3)

(55958, 48)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits unique_customer_id  
0    9821321604.0           5847           9821           58479821  
1    9003429027.0           1516           9003           15169003  
2    8600031276.0           9126           8600           91268600

In [48]:
# Checking for missing values
updated_ro_data.isnull().sum()

Unnamed: 0                        0
VIN                               0
NAME                              1
MODEL                             0
FUEL_TYPE                         0
EMAIL_ID                         35
ADDRESS                        1823
REGISTRATION_NO                  40
PURCHASE_DATE                     4
PREFERRED_DEALER                  0
REPM_CUST_NO                      0
SPEC_AUTO_TRANS_FALG          55958
SPEC_ENGN_CPCTY               55958
MODEL_CODE                        0
CITY_NAME                       541
STATE_NAME                      289
CUST_PIN_CODE                   326
SERVICE_DLR_NO                    0
SALES_DLR_NO                      0
WORK_TYPE                         0
VRNT_CODE                         0
RO_DTIME                          0
RO_NO                             0
MOBILE_NO                         4
OFFICE_NO                     54985
RESIDENCE_NO                  55204
MILEAGE                           0
CUST_DATE_OF_BIRTH          

In [49]:
# Checking the unique value count for this unique customer id
updated_ro_data['unique_customer_id'].nunique()

41569

In [51]:
# Checking unique VIN count
updated_ro_data['VIN'].nunique()

55958

In [50]:
# Checking unique Mob no count
updated_ro_data['Verified_mob_no'].nunique()

32536

In [52]:
55958 - 32536

23422

### Adding the complaint Yes/No column to the RO Data

In [53]:
# Loading the complaint data
complaint = pd.read_excel("Complaints_data_YTD-Big_Data.xltm")
print(complaint.shape)
complaint.head(3)

(20226, 73)


SR Number    Area Manager     Region  Status    Customer Name Escalated  \
0  1-3907916904    Ajay Gokapai    South 2  Closed             Ravi       NaN   
1  1-3907489432    Swarnav Bose     East 2  Closed  HARI PRASAD RAY       NaN   
2  1-3907920284  Saurabh Mishra  Central 2  Closed     MUNESH KUMAR       NaN   

           Opened Date SR Opened Month  Complaint Pending For Days#  \
0  2021-01-01 08:00:57         January                          NaN   
1  2021-01-01 08:04:33         January                          NaN   
2  2021-01-01 08:15:38         January                          NaN   

                  Type SR Exclude Flag                       Area  \
0  Complaint - Service               N  Quality of repair related   
1  Complaint - Service               N                     Others   
2  Complaint - Service               Y  Quality of repair related   

                               Sub-Area  \
0  Poor washing, cleaning, final finish   
1            Car damaged during repairs   
2             Poor & incomplete repairs   

                                         Description theft  AVN  \
0  [CITY] I0587 [REGISTRATION] TS08GV6279 [VIN] M...   NaN  NaN   
1  GRAND I10,CUSTOMER COMPLAINT THAT 1.HE GAVE HI...   NaN  NaN   
2  [CITY] 512 [REGISTRATION] UP78CK1429 [VIN]  Ka...   NaN  NaN   

                                          HO Remarks  \
0  CCSUSER:1/4/2021 11:14:44 : RO Ack Done :Mon 1...   
1  CCSUSER:1/4/2021 17:17:50 : RO Ack Done :Mon 1...   
2  BHAWNA:1/12/2021 15:21:53 : rat bite concern. ...   

                                          RO Remarks DLR SR Exclude Flag  \
0  SOUTH2SERVICE:1/4/2021 12:46:33 : Washing impr...                   N   
1  EAST2SERVICE:1/4/2021 11:59:53 : Car Damaged i...                   N   
2                                                NaN                   Y   

                                        GDMS Remarks  \
0   : Appolozised cust. for inconvenience happen ...   
1   : FI NOT DONE : CUSTOMER VISIT OUR WORKSHOP T...   
2                            : work done : work done   

                          Dealer Super Dealer # Dealer# Dealer Ack Flag  \
0  RVAN VENTURES PRIVATE LIMITED          S5230   S5230             NaN   
1        Frostees India Pvt. Ltd          E4200   E4705             NaN   
2               TIRUPATI HYUNDAI          N3215   N3417             NaN   

  Dealer Ack Date  Acknowledgement Time SR Closed Month          Closed Date  \
0             NaN                   NaN         January  2021-02-01 19:17:20   
1             NaN                   NaN         January  2021-04-01 16:43:51   
2             NaN                   NaN         January  2021-02-01 13:41:33   

   Closing Time                VIN Registration #   Mileage VHCL Sale Date  \
0           1.0  MALA851CLKM071385     TS08GV6279    9741.0     12/30/2019   
1           3.0  MALB251CLLM111187     AS01EQ8867   21108.0     12/21/2020   
2           1.0  MALBB51BLBM310545     UP78CK1429  107907.0     07/27/2011   

                 Closing Doc Type  \
0  Home Visit & Satisfaction note   
1              Satisfaction Notes   
2              Satisfaction Notes   

                                         Product Source             Origin  \
0                 Grand i10 Grand i10 Sportz 1.2    Web         Mobile App   
1  Grand i10 NIOS Grand i10 NIOS 1.2MT Corporate  Phone        Call Center   
2                              i20 ASTA 1.2 BSIV    Web  Corporate Website   

  ASM/RPSM Remarks HO Attrib1 Last RO Advisor Name Reassign Date Reopen Date  \
0              NaN        NaN           G.SRIKANTH           NaN         NaN   
1              NaN        NaN          RINKU SARMA           NaN         NaN   
2              NaN        NaN       VIJAY TRIPATHI           NaN         NaN   

   Reopen Count  Repeated Count  Model Year  Last RO Type  RO Attribute1 Type  \
0             0             NaN      2019.0  Free Service  Mechanical Process   
1             0             0.0      2020.

In [54]:
# Checking for null values
complaint.isnull().sum()

SR Number                          0
Area Manager                      54
Region                             3
Status                             0
Customer Name                      5
Escalated                      18241
Opened Date                        0
SR Opened Month                    0
Complaint Pending For Days#    19873
Type                               0
SR Exclude Flag                    0
Area                               0
Sub-Area                           0
Description                        1
theft                          20210
AVN                            19620
HO Remarks                       309
RO Remarks                      8892
DLR SR Exclude Flag                0
GDMS Remarks                    2505
Dealer                             3
Super Dealer #                     3
Dealer#                            3
Dealer Ack Flag                15099
Dealer Ack Date                15099
Acknowledgement Time           15099
SR Closed Month                  354
C

In [55]:
# Adding a complaint yes/no column, for all the rows in this data frame the value will be 1
complaint['Complaint'] = 'Yes'
print(complaint.shape)
complaint.head(3)

(20226, 74)


SR Number    Area Manager     Region  Status    Customer Name Escalated  \
0  1-3907916904    Ajay Gokapai    South 2  Closed             Ravi       NaN   
1  1-3907489432    Swarnav Bose     East 2  Closed  HARI PRASAD RAY       NaN   
2  1-3907920284  Saurabh Mishra  Central 2  Closed     MUNESH KUMAR       NaN   

           Opened Date SR Opened Month  Complaint Pending For Days#  \
0  2021-01-01 08:00:57         January                          NaN   
1  2021-01-01 08:04:33         January                          NaN   
2  2021-01-01 08:15:38         January                          NaN   

                  Type SR Exclude Flag                       Area  \
0  Complaint - Service               N  Quality of repair related   
1  Complaint - Service               N                     Others   
2  Complaint - Service               Y  Quality of repair related   

                               Sub-Area  \
0  Poor washing, cleaning, final finish   
1            Car damaged during repairs   
2             Poor & incomplete repairs   

                                         Description theft  AVN  \
0  [CITY] I0587 [REGISTRATION] TS08GV6279 [VIN] M...   NaN  NaN   
1  GRAND I10,CUSTOMER COMPLAINT THAT 1.HE GAVE HI...   NaN  NaN   
2  [CITY] 512 [REGISTRATION] UP78CK1429 [VIN]  Ka...   NaN  NaN   

                                          HO Remarks  \
0  CCSUSER:1/4/2021 11:14:44 : RO Ack Done :Mon 1...   
1  CCSUSER:1/4/2021 17:17:50 : RO Ack Done :Mon 1...   
2  BHAWNA:1/12/2021 15:21:53 : rat bite concern. ...   

                                          RO Remarks DLR SR Exclude Flag  \
0  SOUTH2SERVICE:1/4/2021 12:46:33 : Washing impr...                   N   
1  EAST2SERVICE:1/4/2021 11:59:53 : Car Damaged i...                   N   
2                                                NaN                   Y   

                                        GDMS Remarks  \
0   : Appolozised cust. for inconvenience happen ...   
1   : FI NOT DONE : CUSTOMER VISIT OUR WORKSHOP T...   
2                            : work done : work done   

                          Dealer Super Dealer # Dealer# Dealer Ack Flag  \
0  RVAN VENTURES PRIVATE LIMITED          S5230   S5230             NaN   
1        Frostees India Pvt. Ltd          E4200   E4705             NaN   
2               TIRUPATI HYUNDAI          N3215   N3417             NaN   

  Dealer Ack Date  Acknowledgement Time SR Closed Month          Closed Date  \
0             NaN                   NaN         January  2021-02-01 19:17:20   
1             NaN                   NaN         January  2021-04-01 16:43:51   
2             NaN                   NaN         January  2021-02-01 13:41:33   

   Closing Time                VIN Registration #   Mileage VHCL Sale Date  \
0           1.0  MALA851CLKM071385     TS08GV6279    9741.0     12/30/2019   
1           3.0  MALB251CLLM111187     AS01EQ8867   21108.0     12/21/2020   
2           1.0  MALBB51BLBM310545     UP78CK1429  107907.0     07/27/2011   

                 Closing Doc Type  \
0  Home Visit & Satisfaction note   
1              Satisfaction Notes   
2              Satisfaction Notes   

                                         Product Source             Origin  \
0                 Grand i10 Grand i10 Sportz 1.2    Web         Mobile App   
1  Grand i10 NIOS Grand i10 NIOS 1.2MT Corporate  Phone        Call Center   
2                              i20 ASTA 1.2 BSIV    Web  Corporate Website   

  ASM/RPSM Remarks HO Attrib1 Last RO Advisor Name Reassign Date Reopen Date  \
0              NaN        NaN           G.SRIKANTH           NaN         NaN   
1              NaN        NaN          RINKU SARMA           NaN         NaN   
2              NaN        NaN       VIJAY TRIPATHI           NaN         NaN   

   Reopen Count  Repeated Count  Model Year  Last RO Type  RO Attribute1 Type  \
0             0             NaN      2019.0  Free Service  Mechanical Process   
1             0             0.0      2020.

In [56]:
# Adding the complaint target variable to RO Data
final_data_with_target_variable = pd.merge(updated_ro_data, complaint[['VIN', 'Complaint']], how = 'left', left_on = 'VIN', right_on = 'VIN')
print(final_data_with_target_variable.shape)
final_data_with_target_variable.head(3)

(56835, 49)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits unique_customer_id Complaint  
0    9821321604.0           5847           9821           58479821       NaN  
1    9003429027.0           1516           9003           15169003       NaN  
2    8600031276.0           9126           8600           91268600       NaN

In [57]:
56835-55958

877

In [58]:
# Removing duplicates
final_data_with_target_variable.drop_duplicates(subset = ['VIN'], keep= 'first')
print(final_data_with_target_variable.shape)
final_data_with_target_variable.head(3)

(56835, 49)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits unique_customer_id Complaint  
0    9821321604.0           5847           9821           58479821       NaN  
1    9003429027.0           1516           9003           15169003       NaN  
2    8600031276.0           9126           8600           91268600       NaN

In [60]:
# Filling the Complaint column missing values with 'No'
final_data_with_target_variable['Complaint'].fillna('No', inplace=True)
print(final_data_with_target_variable.shape)
final_data_with_target_variable.head(3)

(56835, 49)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits unique_customer_id Complaint  
0    9821321604.0           5847           9821           58479821        No  
1    9003429027.0           1516           9003           15169003        No  
2    8600031276.0           9126           8600           91268600        No

In [61]:
final_data_with_target_variable['Complaint'].value_counts()

No     37306
Yes    19529
Name: Complaint, dtype: int64

Now warranty data, and PSF Data needs to be added in this dataframe.
In case of PSF data, add PSF taken date and then find the difference between RO Dtime and PSF taken date. 
Verify the warranty data.

In [62]:
# Loading the warranty data
warranty = pd.read_excel('VIN_warranty_data.xltm')
print(warranty.shape)
warranty.head(3)

(58908, 9)


VIN   SALE_DATE  BASIC_WRNTY EXTB_VIN  EXTB_WRNTY_NO  \
0  MALC841DMKM127830  20190222.0   20220222.0      NaN            NaN   
1  MALC841CMKM183325  20190929.0   20220929.0      NaN            NaN   
2  MALFC81BLMM186773  20210305.0   20240305.0      NaN            NaN   

  EXTB_SCHM_NO SCHEME_NAME  EXTB_APPL_STRT_DATE  EXTB_APPL_FNSH_DATE  
0          NaN         NaN                  NaN                  NaN  
1          NaN         NaN                  NaN                  NaN  
2          NaN         NaN                  NaN                  NaN

In [63]:
# Removing duplicates from warranty data
warranty.drop_duplicates(subset = ['VIN'], keep = 'first')
print(warranty.shape)
warranty.head(3)

(58908, 9)


VIN   SALE_DATE  BASIC_WRNTY EXTB_VIN  EXTB_WRNTY_NO  \
0  MALC841DMKM127830  20190222.0   20220222.0      NaN            NaN   
1  MALC841CMKM183325  20190929.0   20220929.0      NaN            NaN   
2  MALFC81BLMM186773  20210305.0   20240305.0      NaN            NaN   

  EXTB_SCHM_NO SCHEME_NAME  EXTB_APPL_STRT_DATE  EXTB_APPL_FNSH_DATE  
0          NaN         NaN                  NaN                  NaN  
1          NaN         NaN                  NaN                  NaN  
2          NaN         NaN                  NaN                  NaN

In [64]:
# Merging this data with RO data
master_data = pd.merge(final_data_with_target_variable, warranty[['VIN','SALE_DATE', 'BASIC_WRNTY', 'EXTB_VIN', 'EXTB_WRNTY_NO', 'EXTB_APPL_STRT_DATE', 'EXTB_APPL_FNSH_DATE']], how = 'left', on = 'VIN')
print(master_data.shape)
master_data.head(3)

(58310, 55)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits unique_customer_id Complaint  \
0    9821321604.0           5847           9821           58479821        No   
1    9003429027.0           1516           9003           15169003        No   
2    8600031276.0           9126           8600           91268600        No   

    SALE_DATE  BASIC_WRNTY EXTB_VIN  EXTB_WRNTY_NO  EXTB_APPL_STRT_DATE  \
0  20101203.0   20131203.0      NaN            NaN                  NaN   
1  20110608.0          NaN      NaN            NaN                  NaN   
2  20120331.0   20140331.0  

In [66]:
58310-56835

1475

In [65]:
# Removing duplicates
master_data.drop_duplicates(subset = ['VIN'], keep = 'first')
print(master_data.shape)
master_data.head(3)

(58310, 55)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits unique_customer_id Complaint  \
0    9821321604.0           5847           9821           58479821        No   
1    9003429027.0           1516           9003           15169003        No   
2    8600031276.0           9126           8600           91268600        No   

    SALE_DATE  BASIC_WRNTY EXTB_VIN  EXTB_WRNTY_NO  EXTB_APPL_STRT_DATE  \
0  20101203.0   20131203.0      NaN            NaN                  NaN   
1  20110608.0          NaN      NaN            NaN                  NaN   
2  20120331.0   20140331.0  

In [67]:
# Loading the 3rd day PSF data
psf = pd.read_excel('3rd_Day_PSF_Sample_Data.xltm')
print(psf.shape)
psf.head(3)

(47601, 45)


R1    RNUM FLWUP_CMPN_NO FLWUP_CORP_NO     FLWUP_NO  FLWUP_DATE  \
0  500001  102436             H         A10AA  P2021005038  27/09/2021   
1  500002  102453             H         A10AA  P2021001646  27/09/2021   
2  500003  102474             H         A10AA  P2021003440  29/09/2021   

  FLWUP_CUST_NO FLWUP_DLR_NO          FLWUP_VIN FLWUP_SPC_MSG_DSCTN  \
0   C2021090414        S4237  MALBB51RLCM442869                 NaN   
1   C2020130843        S4238  MALAN51CLDM456558                 NaN   
2   C2021090466        S4240  MALA841CLFM111510                 NaN   

  FLWUP_RO_NO FLWUP_RO_DATE FLWUP_RGSTN_NO  FLWUP_SALE_DATE  FLWUP_RGSTN_DATE  \
0  R202105948    23/09/2021     TN20CY2381       20120721.0        20120720.0   
1  R202101998    22/09/2021     TN41AJ6792       20140110.0               NaN   
2  R202104115    23/09/2021     TN49BD6787       20150818.0        20150908.0   

   FLWUP_LAST_MILG FLWUP_DUE_DATE  FLWUP_DIFF_DUE_DATE FLWUP_SVC_DUE_DATE  \
0           106812     26/09/2021                  -34         26/09/2021   
1            35554     26/09/2021                  -34         26/09/2021   
2           104245     26/09/2021                  -34         26/09/2021   

  FLWUP_CALL_DATE                     FLWUP_DLR_DSCTN FLWUP_VST_TYPE  \
0      27/09/2021             PEEYESYEM HK MOTORS LLP            NaN   
1      27/09/2021  CHANDRA AUTOMOBILE INDIA PVT. LTD.            NaN   
2      29/09/2021              SRI VETRI CARS PVT LTD            NaN   

  FLWUP_VST_TYPE_NM  FLWUP_CALL FLWUP_CALL_NAME     FLWUP_SA_NAME  \
0               NaN           2     3rd day PSF  DHEENADHAYALAN P   
1               NaN           2     3rd day PSF           MANOJ.S   
2               NaN           2     3rd day PSF      MANIKANDAN G   

  FLWUP_BILL_DATE      FLWUP_SV_TYPE FLWUP_STATUS  FLWUP_STATUS_CODE  \
0      23/09/2021       Paid Service    Satisfied               11.0   
1      22/09/2021       Paid Service    Satisfied               11.0   
2      23/09/2021  Accidental Repair    Satisfied               11.0   

                        FLWUP_SRV_RMRKS FLWUP_FUEL_TYPE FLWUP_FUEL_NAME  \
0                             satisfied               D          Diesel   
1                                    10               U          Petrol   
2  steering tight and AC coolining low                U          Petrol   

   FLWUP_RMNDR_CNT  FLWUP_CURR_TYPE  NO_FLWUP_REASON CALL_EMP_NO  \
0              NaN                1              NaN  ES42370140   
1              NaN                1              NaN  ES42380052   
2              NaN                1              NaN  ES42400029   

  CALL_EMP_NO_NM   OB_CALL_NO   SVY_RST_NO  SERVICE_TYPE MODEL_CODE_1  \
0       DEVIKA P  P2021005038  V2109045039           NaN           CN   
1      V.MALATHI  P2021001646  V2109033092           NaN           CM   
2      NIVETHA.G  P2021003440  V2109048680           NaN           XX   

     SVY_DATE    PSF_QUS          PSF_ANS  
0  20210927.0  1,2,3,4,5  Yes,10,10,10,10  
1  20210927.0  1,2,3,4,5  Yes,10,10,10,10  
2  20210929.0  1,2,3,4,5      Yes,9,9,9,9

In [70]:
# Removing duplicates from PSF Data
psf = psf.drop_duplicates(subset = ['FLWUP_VIN'], keep = 'first')
print(psf.shape)

(46812, 45)


For merging the dataset remove duplicates before merging. So that the merged data set doesn't get duplicate rows.

In [71]:
# Taking into consideration only the follow up customer status column and merging with RO Data
updated_master_data = pd.merge(master_data, psf[['FLWUP_VIN','FLWUP_STATUS']], how = 'left', left_on = 'VIN', right_on = 'FLWUP_VIN')
print(updated_master_data.shape)
updated_master_data.head(3)

(58310, 57)


Unnamed: 0                VIN                  NAME           MODEL  \
0           0  KMHSH81XLAU665847        D K ASSOCIATES  SANTA FE (CMi)   
1           1  KMHSH81XLBU741516   PREM DEVA PRASATH.D  SANTA FE (CMi)   
2       47066  MALA151ALCM039126  JASWANT SINGH DODIYA             EON   

  FUEL_TYPE                    EMAIL_ID  \
0         D     dkumars9@rediffmail.com   
1         D  dpremdevaprasath@gmail.com   
2         U      lalipcodiiya@gmail.com   

                                    ADDRESS REGISTRATION_NO  PURCHASE_DATE  \
0                                    MUMBAI      MH04EQ5769     20101203.0   
1  4/2,UPPUTHOTTAM STREET,AVINASHI ROAD ,AN      TN10AC9700     20110608.0   
2                           KASTURBA NAGAR        MP43C5146     20120331.0   

  PREFERRED_DEALER REPM_CUST_NO  SPEC_AUTO_TRANS_FALG  SPEC_ENGN_CPCTY  \
0            W1202  C2017090133                   NaN              NaN   
1            S4407  C2011060072                   NaN              NaN   
2            W5237  C2019052081                   NaN              NaN   

  MODEL_CODE CITY_NAME      STATE_NAME CUST_PIN_CODE SERVICE_DLR_NO  \
0         BM    MUMBAI     Maharashtra        400081          W1202   
1         BM  AVANASHI       Tamilnadu        641653          S4407   
2         IA    Ratlam  MADHYA PRADESH        457001          W5237   

  SALES_DLR_NO WORK_TYPE  VRNT_CODE             RO_DTIME       RO_NO  \
0        W1201        RR        963  2021-08-25 13:52:04  R202116390   
1        S1100        RR        963  2021-09-29 16:16:28  R202104381   
2        W5208        PS       1178  2021-09-13 11:23:59  R202103318   

    MOBILE_NO  OFFICE_NO  RESIDENCE_NO  MILEAGE  CUST_DATE_OF_BIRTH  \
0  9821321604        NaN           NaN    89063                 NaN   
1  9003429027        NaN           NaN   149036                 NaN   
2  8600031276        NaN           NaN    18177                 NaN   

  CUST_GENDER  CUST_AGE   PROFESSION  SERVICE_ADVISOR_DATE_BIRTH  \
0           M       NaN          NaN                         NaN   
1           M       NaN          NaN                         NaN   
2           M       NaN  Businessmen                         NaN   

   SERVICE_ADVISOR_AGE  REPM_ESTMT_AMT REPM_SVC_ADVSR_EMP_NO  \
0                  NaN             0.0                  TEC5   
1                  NaN          2000.0                 KM015   
2                  NaN             0.0            EW52370040   

   REPM_BILL_TOTAL_PART_AMT  REPM_BILL_TOTAL_LABR_AMT  \
0                  50214.94                    6195.0   
1                      0.00                    1888.0   
2                   2376.98                    3380.7   

   REPM_BILL_TOTAL_OTHR_AMT      REPM_CLSE_DTIME INVC_TYPE  \
0                         0  2021-08-25 16:23:04    Normal   
1                         0  2021-09-29 17:15:47    Normal   
2                         0  2021-09-13 17:35:48    Normal   

                            DEF_CODE  \
0  121,A05,B07,BD154,BR001,CL99,S002   
1                           AW25,Z01   
2                                Z01   

                                        Deffered Job       SA_NAME Taxi_flag  \
0  CHECK BRAKE DISC RUSTY/SKIMMING,AC Cooling Les...  KISHOR KADAM        No   
1                                    Wheel Alignment    Karthik K.        No   
2                                                NaN  IBRAHIM KHAN        No   

  Verified_mob_no VIN_end_digits Mob_end_digits unique_customer_id Complaint  \
0    9821321604.0           5847           9821           58479821        No   
1    9003429027.0           1516           9003           15169003        No   
2    8600031276.0           9126           8600           91268600        No   

    SALE_DATE  BASIC_WRNTY EXTB_VIN  EXTB_WRNTY_NO  EXTB_APPL_STRT_DATE  \
0  20101203.0   20131203.0      NaN            NaN                  NaN   
1  20110608.0          NaN      NaN            NaN                  NaN   
2  20120331.0   20140331.0  

In [74]:
# Saving sample master data to csv
updated_master_data.to_csv('Sample_Merged_Dataset.csv', encoding = 'utf-8')

In [72]:
# Checking missing values 
updated_master_data.isnull().sum()

Unnamed: 0                        0
VIN                               0
NAME                              1
MODEL                             0
FUEL_TYPE                         0
EMAIL_ID                         36
ADDRESS                        1864
REGISTRATION_NO                  41
PURCHASE_DATE                     4
PREFERRED_DEALER                  0
REPM_CUST_NO                      0
SPEC_AUTO_TRANS_FALG          58310
SPEC_ENGN_CPCTY               58310
MODEL_CODE                        0
CITY_NAME                       576
STATE_NAME                      307
CUST_PIN_CODE                   346
SERVICE_DLR_NO                    0
SALES_DLR_NO                      0
WORK_TYPE                         0
VRNT_CODE                         0
RO_DTIME                          0
RO_NO                             0
MOBILE_NO                         5
OFFICE_NO                     57283
RESIDENCE_NO                  57521
MILEAGE                           0
CUST_DATE_OF_BIRTH          

In [73]:
58310-54692
# Follow up status from PSF data is only available for 3618 records

3618